peer/integrationtest/tchannel_roundrobin_test.go

```diff
@@ -0,0 +1,163 @@
+// THE SOFTWARE.
+
+package integrationtest
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"reflect"
+	"testing"
+	"time"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/encoding/raw"
+	peerbind "go.uber.org/yarpc/peer"
+	"go.uber.org/yarpc/peer/hostport"
+	"go.uber.org/yarpc/peer/roundrobin"
+	"go.uber.org/yarpc/transport/tchannel"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestTChannelWithRoundRobin verifies that TChannel appropriately notifies all
+// subscribed peer lists when peers become available and unavailable.
+// It does so by constructing a round robin peer list backed by the TChannel transport,
+// communicating to three servers. One will always work. One will go down
+// temporarily. One will be a bogus TCP port that never completes a TChannel
+// handshake.
+func TestTChannelWithRoundRobin(t *testing.T) {
+	// Setup
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+
+	permanent, permanentAddr := server(t, "")
+	defer permanent.Stop()
+
+	temporary, temporaryAddr := server(t, "")
+	defer temporary.Stop()
+
+	l, err := net.Listen("tcp", ":0")
+	require.NoError(t, err, "listen for bogus server")
+	invalidAddr := l.Addr().String()
+	defer l.Close()
+
+	// Construct a client with a bank of peers. We will keep one running all
+	// the time. We'll shut one down temporarily. One will be invalid.
+	// The round robin peer list should only choose peers that have
+	// successfully connected.
+	client, c := client(t, []string{
+		permanentAddr,
+		temporaryAddr,
+		invalidAddr,
+	})
+	defer client.Stop()
+
+	// All requests should succeed. The invalid peer never enters the rotation.
+	blast(ctx, t, c)
+
+	// Shut down one task in the peer list.
+	temporary.Stop()
+	// One of these requests may fail since one of the peers has gone down but
+	// the TChannel transport will not know until a request is attempted.
+	call(ctx, c)
+	call(ctx, c)
+	// All subsequent should succeed since the peer should be removed on
+	// connection fail.
+	blast(ctx, t, c)
+
+	// Restore the server on the temporary port.
+	restored, _ := server(t, temporaryAddr)
+	defer restored.Stop()
+	blast(ctx, t, c)
+}
+
+func blast(ctx context.Context, t *testing.T, c raw.Client) {
+	assert.NoError(t, call(ctx, c))
+	assert.NoError(t, call(ctx, c))
+	assert.NoError(t, call(ctx, c))
+	assert.NoError(t, call(ctx, c))
+	assert.NoError(t, call(ctx, c))
+}
+
+func call(ctx context.Context, c raw.Client) error {
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+	res, err := c.Call(ctx, "echo", []byte("hello"))
+	if err != nil {
+		return err
+	}
+	if !reflect.DeepEqual(res, []byte("hello")) {
+		return fmt.Errorf("unexpected response %+v", res)
+	}
+	return nil
+}
+
+func client(t *testing.T, addrs []string) (*yarpc.Dispatcher, raw.Client) {
```

Comment:
ya note comment on `call` - just be consistent, pass `testing.T` to call as well?